ユーザー管理機能の実装

tomo1107 opened the pull request:
# What
## 実装内容
- deviseを用いたユーザーの新規登録・ログイン・ログアウト機能の実装
- モデルの単体テストの実装
- トップページへの[新規登録・ログイン・ログアウト]ボタンの実装
- ログイン・ログアウト状態に合わせてトップページのヘッダーが変化する

## 動作(Gyazo画像)
- 正しく情報を記入すると、ユーザーの新規登録ができる動画
https://gyazo.com/ff276c29d21706dc3b655491489c6f17
- 入力に問題のある状態ではユーザーの新規登録はできず、エラーメッセージが出力される動画
https://gyazo.com/a26128a807767c675c18b515bf698e3e
- 正しく情報を記入すると、ユーザーのログインができる動画
https://gyazo.com/fdbbddfb41f10ea3e2df9740935f136d
- 入力に問題のある状態ではユーザーのログインはできず、エラーメッセージが出力される動画　
https://gyazo.com/640d85dc8c1052d73daabc6123a3ffc8
- ログアウトに成功する動画
https://gyazo.com/3b3a732c883a1f6526f07eab76ae8c3b
- テスト成功の画像
https://gyazo.com/febf9ff9c47fbd90f6da46b2bd83357d

# Why
- ユーザーの新規登録・ログイン・ログアウト機能が正しく実装されているかを確認頂きたい。
- モデルの単体テストが正しく実装されているかを確認頂きたい。

tomo1107 committed:
新規登録・ログイン・ログアウト機能の実装

tomo1107 committed:
モデル単体テストの実装

yutomentors commented on app/models/user.rb:
```diff
@@ -0,0 +1,25 @@
+  FULL_WIDTH_KATAKANA_REGEX = /\A[ァ-ヶー－]+\z/.freeze     #全角カタカナのみ
+  PASSWORD_REGEX = /\A(?=.*?[a-z])(?=.*?[\d])[a-z\d]+\z/i.freeze     #英数字混合
+  
+  validates_format_of :password, with: PASSWORD_REGEX, message: 'Include both letters and numbers' 
+  validates :nickname, presence: true
+  validates :first_name, presence: true
+  validates_format_of :first_name, with: FULL_WIDTH_REGEX, message: 'Full-width characters' 
+  validates :last_name, presence: true
+  validates_format_of :last_name, with: FULL_WIDTH_REGEX, message: 'Full-width characters' 
+  validates :kana_first_name, presence: true
+  validates_format_of :kana_first_name, with: FULL_WIDTH_KATAKANA_REGEX, message: 'Full-width katakana characters' 
+  validates :kana_last_name, presence: true
+  validates_format_of :kana_last_name, with: FULL_WIDTH_KATAKANA_REGEX, message: 'Full-width katakana characters' 
+  validates :birthday, presence: true
```
11~21行目についてですが、`presence true`についての項目は一つにまとめて記述してみましょう！

※`with_options presence: true,`のようにひとまとめにすることで可読性向上にも繋がります。

tomo1107 replied:
with_optionsを使用して、まとめれるバリデーションをまとめました。

yutomentors commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,106 @@
+      </h1>
+    </div>
+
+    <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
+    <%= render 'shared/error_messages', model: f.object %>
+    <%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
```
<%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```

こちらのコメントアウトですが、不要であれば削除しましょう。
もし必要であれば、第三者でもわかるように「残す目的」「なぜ必要か」という理由などを、コメントとして書いておくことをおすすめします。

tomo1107 replied:
コメントアウトを削除しました。

yutomentors reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、ご確認をお願いいたします！

修正頂いた後、コミットプッシュを確認いただき、再度レビュー依頼のご提出をお願い致します！

tomo1107 committed:
是正対応1回目

以下の指摘に対して是正しました。
https://github.com/tomo1107/furima-35073/pull/2#pullrequestreview-631133802

tomo1107 commented:
> レビュー依頼ありがとうございます！
> コメントしましたので、ご確認をお願いいたします！
> 
> 修正頂いた後、コミットプッシュを確認いただき、再度レビュー依頼のご提出をお願い致します！

レビュー頂きありがとうございました。
ご指摘に対しての是正を行いましたので、再レビューの程、よろしくお願いします。

- モデルの単体テストの結果です。
https://gyazo.com/f657d6815f5a1b2e1ea2685f4aaaf4c1

TakashiKozuma commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,145 @@
+      @user.birthday = ''
+      @user.valid?
+      expect(@user.errors.full_messages).to include("Birthday can't be blank")
+    end
```
現状だと正常系と異常系のテストコードが混在しています。
以下のようにcontextを使用して正常系と異常系とを分けて記述すると
可読性の高いテストコードとなりますので修正しましょう。

```
context 'ユーザ登録ができる時' do
  it '~~~' do
    ~~~
  end
end

context 'ユーザ登録ができない時' do
  it '~~~' do
    ~~~
  end

  it '~~~' do
    ~~~
  end
end
```

tomo1107 replied:
contextを使用して正常系と異常系とを分けてexampleを記述しました。

TakashiKozuma reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認致しました。
テストコードに関してコメントを追記いたしましたので、ご確認お願いします。

tomo1107 committed:
是正対応2回目

以下のご指摘に対しての是正対応
https://github.com/tomo1107/furima-35073/pull/2#pullrequestreview-631174222

tomo1107 commented:
> レビュー依頼ありがとうございます！
> 修正点確認致しました。
> テストコードに関してコメントを追記いたしましたので、ご確認お願いします。

レビュー頂きありがとうございました。
ご指摘に対しての是正を行いましたので、再レビューの程、よろしくお願いします。
- 是正後の単体テストの結果です。
https://gyazo.com/293b92045cd7bdc6cde31f147917e328

TakashiKozuma commented:
レビュー依頼ありがとうございます。
テストの修正も確認いたしました！
LGTMです！実装お疲れさまでした。

こちらのブランチは忘れずにMergeいただき、ローカルのMasterにPullした後に次の実装のためのブランチを切ってから次にお進み下さい。
